engine/apps/mattermost/models/message.py

```diff
@@ -0,0 +1,33 @@
+from django.db import models
+
+from apps.alerts.models import AlertGroup
+from apps.mattermost.client import MattermostPost
+
+
+class MattermostMessage(models.Model):
+    (
+        ALERT_GROUP_MESSAGE,
+        LOG_MESSAGE,
+    ) = range(2)
+
+    MATTERMOST_MESSAGE_CHOICES = ((ALERT_GROUP_MESSAGE, "Alert group message"), (LOG_MESSAGE, "Log message"))
+
+    post_id = models.CharField(max_length=100)
+
+    channel_id = models.CharField(max_length=100)
+
+    message_type = models.IntegerField(choices=MATTERMOST_MESSAGE_CHOICES)
+
+    alert_group = models.ForeignKey(
+        "alerts.AlertGroup",
+        on_delete=models.CASCADE,
+        related_name="mattermost_messages",
+    )
```

Comment:
I think we should have some unique condition (+ use `get_or_create` in the task) between `alert_group`, `channel_id` and `message_type` so some delay task/retries do not create multiple notifications for the same alert group (this happened to me in local testing :-))

Reply:
I am not sure if i understand completely. When we receive the signal for an alert group we enqueue a task which is like a 1:1 mapping for the task and alert group. In case of failures or retry won't it be happening for that alert group only ? 

Also `alert_group, channel_id and message_type` does not uniquely identify the message since for messages in the same thread share a common value for the above combination.

I am not able to see how multiple notifications gets triggered for the same alert group. Can you help me see where i am missing ?

Reply:
> I am not sure if i understand completely. When we receive the signal for an alert group we enqueue a task which is like a 1:1 mapping for the task and alert group. In case of failures or retry won't it be happening for that alert group only ?

Yeah, but it seems there is some retry condition or update that could enqueue multiple tasks for the same alert group, leading to duplicated notifications (need to take a closer look to check what the context was).

> Also `alert_group, channel_id and message_type` does not uniquely identify the message since for messages in the same thread share a common value for the above combination.

But we don't need to track every message in a thread, right? Only the root (alert group notification) message, which should be only one per alert group?

> I am not able to see how multiple notifications gets triggered for the same alert group. Can you help me see where i am missing ?

As mentioned above, not really sure what the specific conditions triggering the issue were (I was missing some Mattermost secret in my setup first, when set, I got duplicated messages).

Reply:
I was not able to reproduce the scenario in my local but I have added the condition to prevent duplicate events and also fixed the linting and test errors.

Reply:
To be sure, I don't see a unique constraint for the model, should there be one?

Reply:
I think i missed it, I have added a conditional unique constraint plus we have foreign key reference for alert group which we mostly use for fetching the messages.